### PR TITLE
Add Delta Tailoring Files to the Build System

### PR DESCRIPTION
**Pull request by Mab879**
#### Description:

Builds the delta tailoring files for RHEL7 and RHEL8.

#### Rationale:

So that we can releases these files to downstream.

### CODE REVIEW

File: cmake/SSGCommon.cmake [hunk 1]
```diff
@@ -761,6 +761,32 @@ macro(ssg_make_all_tables PRODUCT)
     )
 endmacro()
 
+macro(ssg_build_rule_dir_json)
+    add_custom_command(
+        OUTPUT "${CMAKE_BINARY_DIR}/rule_dirs.json"
+        COMMAND env "PYTHONPATH=$ENV{PYTHONPATH}" "${PYTHON_EXECUTABLE}" "${CMAKE_SOURCE_DIR}/utils/rule_dir_json.py" --output "${CMAKE_BINARY_DIR}/rule_dirs.json" --root "${CMAKE_SOURCE_DIR}"
```
Comment (yuumasato) [inline cmake/SSGCommon.cmake:767]: @jan-cerny @matejak Thoughts on generating the `build/rule_dirs.json` by default?

I know there were discussions about this in the past, but couldn't find them.

File: cmake/SSGCommon.cmake [hunk 2]
```diff
@@ -761,6 +761,32 @@ macro(ssg_make_all_tables PRODUCT)
...
+endmacro()
+
+macro(ssg_build_disa_delta PRODUCT PROFILE)
+    file(GLOB DISA_SCAP_REF "${SSG_SHARED_REFS}/disa-stig-${PRODUCT}-v[0-9]*r[0-9]*-xccdf-scap.xml")
+        add_custom_command(
+            OUTPUT "${CMAKE_BINARY_DIR}/${PRODUCT}_{$PROFILE}_delta_tailoring.xml"
+            COMMAND env "PYTHONPATH=$ENV{PYTHONPATH}" "${PYTHON_EXECUTABLE}" "${CMAKE_SOURCE_DIR}/utils/create_scap_delta_tailoring.py" --root "${CMAKE_SOURCE_DIR}" --product "${PRODUCT}" --manual "${DISA_SCAP_REF}" --json "${CMAKE_BINARY_DIR}/rule_dirs.json" --profile "${PROFILE}" --reference "stigid" --output "${CMAKE_BINARY_DIR}/${PRODUCT}_${PROFILE}_delta_tailoring.xml"
+            DEPENDS "${CMAKE_BINARY_DIR}/ssg-${PRODUCT}-ds.xml"
+            DEPENDS "${CMAKE_BINARY_DIR}/rule_dirs.json"
+            COMMENT "[${PRODUCT}-generate-ssg-delta] generating disa tailering file"
```
Comment (yuumasato) [inline cmake/SSGCommon.cmake:782]: Typo: `tailering`

File: cmake/SSGCommon.cmake [hunk 3]
```diff
@@ -761,6 +761,32 @@ macro(ssg_make_all_tables PRODUCT)
...
+        COMMENT "[rule_dir_json] generating rule_dirs.json"
+    )
+    add_custom_target("generate-ssg-rule_dir_json"
+        DEPENDS "${CMAKE_BINARY_DIR}/rule_dirs.json"
+    )
+endmacro()
+
+macro(ssg_build_disa_delta PRODUCT PROFILE)
+    file(GLOB DISA_SCAP_REF "${SSG_SHARED_REFS}/disa-stig-${PRODUCT}-v[0-9]*r[0-9]*-xccdf-scap.xml")
+        add_custom_command(
```
Comment (yuumasato) [inline cmake/SSGCommon.cmake:777]: Indentantion seems to be off.

File: cmake/SSGCommon.cmake [hunk 4]
```diff
@@ -761,6 +761,32 @@ macro(ssg_make_all_tables PRODUCT)
...
+            OUTPUT "${CMAKE_BINARY_DIR}/${PRODUCT}_{$PROFILE}_delta_tailoring.xml"
+            COMMAND env "PYTHONPATH=$ENV{PYTHONPATH}" "${PYTHON_EXECUTABLE}" "${CMAKE_SOURCE_DIR}/utils/create_scap_delta_tailoring.py" --root "${CMAKE_SOURCE_DIR}" --product "${PRODUCT}" --manual "${DISA_SCAP_REF}" --json "${CMAKE_BINARY_DIR}/rule_dirs.json" --profile "${PROFILE}" --reference "stigid" --output "${CMAKE_BINARY_DIR}/${PRODUCT}_${PROFILE}_delta_tailoring.xml"
```
Comment (yuumasato) [inline cmake/SSGCommon.cmake:779]: How about outputting the tailoring to `${CMAKE_BINARY_DIR}/${PRODUCT}/tailoring/${PRODUCT}_${PROFILE}_delta_tailoring.xml` ?

File: cmake/SSGCommon.cmake [hunk 5]
```diff
@@ -761,6 +761,32 @@ macro(ssg_make_all_tables PRODUCT)
...
+            DEPENDS "${CMAKE_BINARY_DIR}/ssg-${PRODUCT}-ds.xml"
+            DEPENDS "${CMAKE_BINARY_DIR}/rule_dirs.json"
+            COMMENT "[${PRODUCT}-generate-ssg-delta] generating disa tailering file"
+         )
+
+        add_custom_target(generate-ssg-delta-${PRODUCT}-${PROFILE}
+            DEPENDS "${CMAKE_BINARY_DIR}/${PRODUCT}_{$PROFILE}_delta_tailoring.xml"
```
Comment (yuumasato) [inline cmake/SSGCommon.cmake:786]: There is a typo (`{$PROFILE}`) which causes `create_scap_detal_tailoring.py` to run everytime.

**Review by yuumasato — CHANGES_REQUESTED** [review]
You also need to add `install` commands so that the tailoring file is moved to its directory during package install.
Check out https://github.com/ComplianceAsCode/content/blob/master/CMakeLists.txt#L28 for inspiration.
I'm thinking a good path might be: `"${CMAKE_INSTALL_DATADIR}/scap-security-guide/tailoring"`

I'm also thinking that the whole output of `create_scap_delta_tailoring.py` is unnecessary when it is plugged into the build system. Maybe add an option to make it silent?

File: cmake/SSGCommon.cmake (hunk 1 above)
Comment (jan-cerny) [reply cmake/SSGCommon.cmake:767]: My concern would be that the `utils/rule_dir_json.py` is very slow, ~ 12 s on my beef laptop.

File: cmake/SSGCommon.cmake [hunk 6]
```diff
@@ -761,6 +761,36 @@ macro(ssg_make_all_tables PRODUCT)
...
+            DEPENDS "${CMAKE_BINARY_DIR}/rule_dirs.json"
+            DEPENDS "${PRODUCT}-content"
+            COMMENT "[${PRODUCT}-generate-ssg-delta] generating disa tailoring file"
+         )
+
+        add_custom_target(generate-ssg-delta-${PRODUCT}-${PROFILE}
+            DEPENDS "${CMAKE_BINARY_DIR}/${PRODUCT}/tailoring/${PRODUCT}_${PROFILE}_delta_tailoring.xml"
+        )
+
+        install(FILES "${CMAKE_BINARY_DIR}/${PRODUCT}/tailoring/${PRODUCT}_{$PROFILE}_delta_tailoring.xml"
```
Comment (yuumasato) [inline cmake/SSGCommon.cmake:790]: Typo again, :) `{$PROFILE}`

File: cmake/SSGCommon.cmake [hunk 7]
```diff
@@ -761,6 +761,36 @@ macro(ssg_make_all_tables PRODUCT)
...
+                DESTINATION SSG_TAILORING_INSTALL_DIR)
```
Comment (yuumasato) [inline cmake/SSGCommon.cmake:791]: Expand the variable `SSG_TAILORING_INSTALL_DIR`, :)

`- Installing: /tmp/SSG_TAILORING_INSTALL_DIR/rhel8_stig_delta_tailoring.xml`

File: cmake/SSGCommon.cmake (hunk 1 above)
Comment (yuumasato) [reply cmake/SSGCommon.cmake:767]: The `rule_dirs.json` is used to collect the Ids of rules applicable to the product:
https://github.com/ComplianceAsCode/content/blob/28a5fb54a949065fbed57a9c6e39e5ddc931961a/utils/create_scap_delta_tailoring.py#L56

But as we are plugging the script into the build system, we could use the data that is available as part of the build pipeline. So collecting the Ids of rules applicable to the product can done by listing the files in `build/${PRODUCT}/rules`.

But I also see the value in being able to run `create_scap_delta_tailoring.py` without doing the whole build.
So the script could either receive a `rule_dir_json.py` or a `--resolved-rules-dir`?

During the build the target `generate-ssg-delta-${PRODUCT}-${PROFILE}` would depend on `${PRODUCT}-compile-all`.

**Review by yuumasato — CHANGES_REQUESTED** [review]

**Review by yuumasato — APPROVED** [review]
LGTM.

**Pull request merged**